File: ProgrammingBasics/30-Lenses-master/JavaScript/5-together.js
"use strict";

const getter = (prop) => (obj) => obj[prop];
const setter = (prop) => (val, obj) => ({ ...obj, [prop]: val });

const view = (lens, obj) => lens.get(obj);
const set = (lens, val, obj) => lens.set(val, obj);
const over = (lens, map, obj) => lens.set(map(lens.get(obj)), obj);

const lens = (getter, setter) => ({
  get: (obj) => getter(obj),
  set: (val, obj) => setter(val, obj),
});

// Usage

const person = {
  name: "Marcus Aurelius",
  city: "Rome",
  born: 121,
};

const nameLens = lens(getter("name"), setter("name"));

console.log("view name:", view(nameLens, person));

console.log("set name:", set(nameLens, "Marcus", person));

const upper = (s) => s.toUpperCase();
console.log("over name:", over(nameLens, upper, person));

File: ProgrammingBasics/30-Lenses-master/Haskell/5-together.hs
import Text.Printf
import Data.Char

data Person = Person
  { _name :: String
  , _city :: String
  , _born :: Int
  } deriving Show
person = Person "Marcus Aurelius" "Rome" 121

data Lens s a = Lens (s -> a) (a -> s -> s)
nameGetter = _name
nameSetter name person = person { _name = name }
nameLens = Lens nameGetter nameSetter
view (Lens g _) = g
set (Lens _ s) = s
over (Lens g s) f obj = s (f . g $ obj) obj

upper = map toUpper
main = do
  printf "view name: %v\n" $ view nameLens person
  printf "set name: %v\n" $ show $ set nameLens "Marcus" person
  printf "over name: %v\n" $ show $ over nameLens upper person
